crates/biome_js_analyze/src/semantic_analyzers/nursery/no_invalid_use_before_declaration.rs

```diff
@@ -136,6 +136,20 @@ impl Rule for NoInvalidUseBeforeDeclaration {
                     // type Y = typeof X;
                     // const X = 0;
                     // ```
+                    && reference
```

Comment:
The issue with this fix is: it doesn't report used before declaraion such as `{ a = b, b = "111" }`.

I wonder if we should rather change the implementation of `biome_js_syntax::binding_ext::AnyJsIdentifierBinding::declaration` and consider `JsObjectBindingPatternShorthandProperty`, `JsObjectBindingPattern*` and `JsArrayBindingPattern*` as declarations. I am not sure of the implication in other part of the codebase. This could solve both #1652 and #1648.

Reply:
Sure, I will look into this after some days, quite busy now.

Reply:
Are you still interested in investigating this? Otherwise, I can take a look :)

Reply:
@Conaclos, I get the idea that we have to include `JsObjectBindingPatternShorthandProperty` as declarations but as you mention I found it hard to handle its implication in other parts of codebase, so you take a look on this. I will follow your work.

Reply:
I proposed a fix in #1800